app/src/main/java/org/stepik/android/presentation/course_list/mapper/CourseListStateMapper.kt

```diff
@@ -0,0 +1,101 @@
+import javax.inject.Inject
+
+class CourseListStateMapper
+@Inject
+constructor() {
+    fun mapToLoadMoreState(courseListState: CourseListView.State.Content): CourseListView.State =
+        CourseListView.State.Content(
+            courseListDataItems = courseListState.courseListDataItems,
+            courseListItems = courseListState.courseListItems + CourseListItem.PlaceHolder
+        )
+
+    fun mapFromLoadMoreToSuccess(state: CourseListView.State, items: PagedList<CourseListItem.Data>): CourseListView.State {
+        if (state !is CourseListView.State.Content) {
+            return state
+        }
+
+        return CourseListView.State.Content(
+            courseListDataItems = state.courseListDataItems + items,
+            courseListItems = state.courseListItems.dropLastWhile(CourseListItem.PlaceHolder::equals) + items
+        )
+    }
+
+    fun mapFromLoadMoreToError(state: CourseListView.State): CourseListView.State {
+        if (state !is CourseListView.State.Content) {
+            return state
+        }
+        return CourseListView.State.Content(
+            courseListDataItems = state.courseListDataItems,
+            courseListItems = state.courseListItems.dropLastWhile(CourseListItem.PlaceHolder::equals)
+        )
+    }
+
+    fun mapToEnrollmentUpdateState(state: CourseListView.State, enrolledCourse: Course): CourseListView.State {
+        if (state !is CourseListView.State.Content) {
+            return state
+        }
+
+        return CourseListView.State.Content(
+            courseListDataItems = PagedList(
+                state.courseListDataItems.map {
+                    if (it.course.id == enrolledCourse.id) it.copy(course = enrolledCourse) else it
+                },
+                state.courseListDataItems.page,
+                state.courseListDataItems.hasNext,
+                state.courseListDataItems.hasPrev
+            ),
+            courseListItems = state.courseListDataItems.map {
+                if (it.course.id == enrolledCourse.id) it.copy(
+                    course = enrolledCourse
+                ) else it
+            }
+        )
+    }
+
+    fun mapToContinueCourseUpdateState(state: CourseListView.State, continuedCourse: Course): CourseListView.State {
+        if (state !is CourseListView.State.Content) {
+            return state
+        }
+
+        val indexOf = state.courseListDataItems.indexOfFirst { it.id == continuedCourse.id }
+
+        val courseListDataItems = state.courseListDataItems.mutate {
+            val courseListItem = removeAt(indexOf)
+            add(0, courseListItem)
+        }
+
+        return CourseListView.State.Content(
+            courseListDataItems = PagedList(
+                courseListDataItems,
+                state.courseListDataItems.page,
+                state.courseListDataItems.hasNext,
+                state.courseListDataItems.hasPrev
+            ),
+            courseListItems = courseListDataItems
```

Comment:
is it true, that if I have not data items in `courseListItems` they will be lost?